tensorflow/compiler/mlir/hlo/lib/Transforms/buffer_packing.cc

```diff
@@ -0,0 +1,490 @@
+      maxUserangeId = std::max(maxUserangeId, lastUse);
+      allocInfos.emplace_back(v, allocUserangeId, firstUse, lastUse,
+                              numSegments, 0, userangeIntervals.getValue());
+    }
+
+    // If the window size is zero we need no sorting anymore.
+    if (windowSize == 0)
+      return maxUserangeId;
+    // Sorts the allocation informations to compute the window id. The window id
+    // is used to blure the userange starting position of an allocation.
+    std::sort(allocInfos.begin(), allocInfos.end(),
+              [](const AllocationInfo &a, const AllocationInfo &b) {
+                return a.allocUserangeId < b.allocUserangeId;
+              });
+
+    // resize window id
+    size_t windowId = 0;
+    size_t lastAllocUserangeId = 0;
+    for (auto &allocationInfo : allocInfos) {
+      if (allocationInfo.allocUserangeId > lastAllocUserangeId + windowSize)
+        ++windowId;
+
+      lastAllocUserangeId = allocationInfo.allocUserangeId;
+      allocationInfo.windowId = windowId;
+    }
+    return maxUserangeId;
+  }
+};
+
+/// Pass to pack buffer together to optimize the memeory consumption and to
+/// save allocation operations. A strategy must be passed as a template
+/// argument.
+class BufferPacking : BufferPlacementTransformationBase {
+
+public:
+  template <typename StrategyT>
+  BufferPacking(Operation *op, StrategyT strategy)
+      : BufferPlacementTransformationBase(op),
+        userangeAnalysis(op, allocs, aliases), dominators(op) {
+    std::vector<PackedBuffer> packedBuffers;
+    strategy.optimze(allocs, userangeAnalysis, packedBuffers);
+
+    for (auto &packedBuffer : packedBuffers) {
+      // Find common dominators.
+      Block *block = findAllocationsDominator(packedBuffer.allocBufferOffsets);
+      // Find alloc position operation.
+      mlir::OpBuilder packBuilder(&(block->front()));
+      auto location = block->front().getLoc();
+      auto memrefType = MemRefType::get({packedBuffer.numSegments},
+                                        packBuilder.getIntegerType(8));
+      Value targetBuffer =
+          packBuilder.create<memref::AllocOp>(location, memrefType);
+
+      for (auto &packInfo : packedBuffer.allocBufferOffsets) {
+        Value currentAlloc = packInfo.source;
+        size_t offset = packInfo.offset;
+        Operation *viewDefOp = currentAlloc.getDefiningOp();
+        Location loc = viewDefOp->getLoc();
+        mlir::OpBuilder viewBuilder(viewDefOp);
+
+        // Create a arithmetic ConstantOp with the aligned offset.
+        Value constantOp = viewBuilder.create<mlir::arith::ConstantOp>(
+            loc, viewBuilder.getIndexType(),
+            viewBuilder.getIntegerAttr(viewBuilder.getIndexType(), offset));
+
+        // Store the operands for the ViewOp.
+        SmallVector<Value, 4> newOperands{targetBuffer};
+        newOperands.push_back(constantOp);
+
+        ShapedType shape = currentAlloc.getType().cast<ShapedType>();
```

Comment:
Why cast to `ShapedType` and not `MemRefType` right away?

Reply:
Okey, I will cast it directly.